llrt_core/src/modules/http/url_search_params.rs

```diff
@@ -1,217 +1,386 @@
+// stringified when added to a URL.
+//
+// ```js
+// const params = new URLSearchParams("foo=bar");
+// const url = new URL("http://github.com");
+// url.search = params; // This works as expected
+// ```
 #[rquickjs::methods(rename_all = "camelCase")]
-impl URLSearchParams {
+impl<'js> URLSearchParams {
     #[qjs(constructor)]
-    pub fn new<'js>(ctx: Ctx<'js>, init: Opt<Value<'js>>) -> Result<Self> {
+    pub fn new(ctx: Ctx<'js>, init: Opt<Value<'js>>) -> Result<Self> {
         if let Some(init) = init.into_inner() {
             if init.is_string() {
-                let string: String = init.into_string().unwrap().to_string().unwrap();
-                return Ok(Self::from_str(&string));
+                let string: String = Coerced::from_js(&ctx, init)?.0;
+                return Ok(Self::from_str(string));
             } else if init.is_array() {
-                let array = init.into_array().unwrap();
-                return Ok(Self::from_array(array));
+                return Self::from_array(&ctx, init.into_array().unwrap());
             } else if init.is_object() {
-                let obj = init.into_object().unwrap();
+                return Self::from_object(&ctx, init.into_object().unwrap());
+            }
+        }
+        let url: Url = "http://example.com".parse().unwrap();
 
-                let iterator = Symbol::iterator(ctx.clone());
+        Ok(URLSearchParams {
+            url: Rc::new(RefCell::new(url)),
+        })
+    }
 
-                if obj.contains_key(iterator)? {
-                    let array_object: Object = ctx.globals().get(PredefinedAtom::Array)?;
-                    let array_from: Function = array_object.get(PredefinedAtom::From)?;
-                    let value: Value = array_from.call((obj,))?;
-                    let array = value.into_array().unwrap();
-                    return Ok(Self::from_array(array));
-                }
+    //
+    // Properties
+    //
 
-                let keys = obj.keys::<String>();
-                let key_len = keys.len();
+    #[qjs(get)]
+    pub fn size(&self) -> usize {
+        self.url.borrow().query_pairs().count()
+    }
 
-                let mut params = Vec::with_capacity(key_len);
+    //
+    // Instance methods
+    //
 
-                for key in keys {
-                    let key = key?;
-                    let val = obj.get::<_, Coerced<String>>(&key)?;
-                    params.push((key, val.to_string()))
+    pub fn append(&mut self, key: Coerced<String>, value: Coerced<String>) {
+        self.url
+            .borrow_mut()
+            .query_pairs_mut()
+            .append_pair(key.as_str(), value.as_str());
+    }
+
+    pub fn delete(&mut self, ctx: Ctx<'js>, key: Coerced<String>, value: Opt<Value<'js>>) {
+        let key = key.0;
+        let value = value.into_inner();
+
+        let new_pairs: Vec<_> = self
+            .url
+            .borrow()
+            .query_pairs()
+            .into_owned()
```

Comment:
Since we're cloning down below in the filter. We can probably avoid this and clone only if the value is not undefined by checking the reference instead.

Reply:
Updated to `filter()` first then `map()` -- `collect` can't convert `Cow<&str>` automatically, thus the need to `map` afaik. But that should remove the unnecessary copying. Thoughts?

Reply:
Exactly, that was the tradeoff with this design I was alluding to in the description. We avoid maintaining derived state and syncing it with the Url between `URL` and `URLSearchParams`, but there's definitely more overhead. I really wish the `Url` crate had implemented their own data structure, ¯\\\_(ツ)\_/¯. If we were to use our own data structure like before, we would still need to sync the params to `Url` at some point. So I've only been able to come up with 2 ideas for it:

Assuming we take advantage of `Url` "quirks" for WHATWG URL compliance...

**Same underlying data structure (what I have now):**

- Good, because there are no sync issues with derived state
- Good, because it uses the native capabilities of `Url` which is _mostly_ WHATWG URL compliant
- Bad, because modifying params requires linear time complexity and a lot of memory allocations

**Derived state, separate data structure for params:**

- Good, because we can use a data structure that allows modifying params in constant time without allocating a bunch of extra memory
- Good, because it could reduce the amount code
- Bad, because we can run into sync issues between `URL` and `URLSearchParams` since they would have different underlying data structures
- Bad, because we still need to copy the params into `Url` when working with the `URL` class (e.g. `url.search()`, `url.searchParams`, `url.toString()`, etc)

Have you been able to come up with any other ideas?

Reply:
I think it's fine to keep your implementation since it's probably more likely we're appending rather than modifying.